traversal/procedure/ProcedureVertex.java

```diff
@@ -157,69 +160,74 @@ public boolean overlaps(Thing other, Traversal.Parameters params) {
-            if (eq.isPresent()) iter = iteratorOfAttributesWithTypes(graphMgr, parameters, eq.get());
+            if (eq.isPresent()) iter = iteratorOfAttributesWithTypes(graphMgr, parameters, eq.get(), order);
             else {
                 if (id().isVariable()) types = types.filter(t -> !t.encoding().equals(ROLE_TYPE));
-                iter = types.mergeMap(t -> graphMgr.data().getReadable(t), ASC);
+                iter = types.mergeMapForwardable(t -> graphMgr.data().getReadable(t, order), order);
             }
 
             if (props().predicates().isEmpty()) return iter;
             else return filterPredicates(mapToAttributes(iter), parameters, eq.orElse(null));
         }
 
-        private Forwardable<? extends ThingVertex, Order.Asc> filterIID(Forwardable<? extends ThingVertex, Order.Asc> iterator,
-                                                                        Traversal.Parameters parameters) {
+        private <ORDER extends Order> Forwardable<? extends ThingVertex, ORDER> filterIID(Forwardable<? extends ThingVertex, ORDER> iterator,
+                                                                                          Traversal.Parameters parameters) {
             assert parameters.getIID(id().asVariable()) != null;
             return iterator.filter(v -> v.iid().equals(parameters.getIID(id().asVariable())));
         }
 
-        private Forwardable<? extends ThingVertex, Order.Asc> filterTypes(Forwardable<? extends ThingVertex, Order.Asc> iterator) {
+        private <ORDER extends Order> Forwardable<? extends ThingVertex, ORDER> filterTypes(Forwardable<? extends ThingVertex, ORDER> iterator) {
             return iterator.filter(v -> props().types().contains(v.type().properLabel()));
         }
 
-        Forwardable<? extends AttributeVertex<?>, Order.Asc> filterPredicates(Forwardable<? extends AttributeVertex<?>, Order.Asc> iterator,
-                                                                              Traversal.Parameters parameters) {
+        <ORDER extends Order> Forwardable<? extends AttributeVertex<?>, ORDER> filterPredicates(Forwardable<? extends AttributeVertex<?>, ORDER> iterator,
+                                                                                                Traversal.Parameters parameters) {
             return filterPredicates(iterator, parameters, null);
         }
 
-        Forwardable<? extends AttributeVertex<?>, Order.Asc> filterPredicates(Forwardable<? extends AttributeVertex<?>, Order.Asc> iterator,
-                                                                              Traversal.Parameters parameters,
-                                                                              @Nullable Predicate.Value<?> exclude) {
+        <ORDER extends Order> Forwardable<? extends AttributeVertex<?>, ORDER> filterPredicates(Forwardable<? extends AttributeVertex<?>, ORDER> iterator,
```

Comment:
these methods are all parametrised by the order, which is a typing guarantee that we always maintain the ordering of the iterator we receive/return or meet the ordering required indicated by the arguments